fix(checkbox): usage update (#1120)
carbon-design-system/carbon-website: src/pages/components/checkbox/usage.mdx

---
title: Checkbox
description: Checkboxes are used when there are one or more items to select in a list.
tabs: ["Code", "Usage", "Style", "Accessibility"]
---

<PageDescription>

> Checkboxes are used when there are one or more items to select in a list.

Author: ooh. yeah I think they should be different, thank you!

> Checkboxes are used when there are one or more items to select in a list.

Reviewer: I'm working on something for you, @laurenmrice !

> Checkboxes are used when there are one or more items to select in a list.

Author: thank you!!

> Checkboxes are used when there are one or more items to select in a list.

Author: fixed 👍🏻

</PageDescription>

<AnchorLinks>

<AnchorLink>Overview</AnchorLink>
<AnchorLink>Formatting</AnchorLink>
<AnchorLink>Behaviors</AnchorLink>
<AnchorLink>Content</AnchorLink>
<AnchorLink>Related</AnchorLink>
<AnchorLink>Feedback</AnchorLink>
<AnchorLink>References</AnchorLink>

</AnchorLinks>

## Overview

Checkboxes are used when there are one or more items to select in a list. Each checkbox works independently from other checkboxes in the list, therefore checking an additional box does not affect any other selections.

### When to use

#### Forms

Can be used in forms on a full page, in modals, or on side panels.

#### Filtering and batch action

> Used to filter data either on a page, in a menu, or within a component. Checkboxes are found in our [Multiselect dropdown](components/dropdown/code) to select multiple items and in [Data table](components/data-table/code) for batch editing purposes.

Reviewer 2: Let me know if this would be confusing or not, but on the development side a `Checkbox` component would correspond to this definition from WAI ARIA:

> Dual-state: The most common type of checkbox, it allows the user to toggle between two choices -- checked and not checked.

For selection in a component, like for multi-select, they are styled to show that an option is selected but unfortunately we don't use `Checkbox` in this situation.

The difference is that a checkbox corresponds to the following input in HTML:

```html
<label for="checkbox">This is a checkbox</label>
<input id="checkbox" type="checkbox">
```

Since `input`'s are interactive, they unfortunately aren't able to be used in things like multi-select.

Author: so this is not a checkbox situation for the multiselect dropdown. is that also true for the selectable rows in a data table?

Reviewer 2: I believe for rows in data table we still use a checkbox! This is definitely controversial but seems fair to have it in here since there is no official solution for row selection in a table right now so really nothing works 🤦 lol

---
title: Checkbox

> description: Checkboxes are used when there are multiple items to select in a list. Users can select zero, one, or multiple items.

Reviewer 3: should we just say users can select any number of items?

![Anatomy of a group of checkboxes.](images/checkbox-usage-3.png)

</Column>
</Row>

1. **Group label (optional):** Communicates what needs to be selected below.
2. **Checkbox input:** A checkbox input indicating the appropriate state. By default it is unselected.
3. **Checkbox label:** Describes the information you want to select or unselect.

### Alignment

Checkbox labels fall to the right of their inputs. If there is a checkbox grouping, they can be laid out vertically or horizontally depending on the use case and the structure of the UI. When possible, arrange the checkbox and radio button groups vertically for easier reading.

<Row>
<Column colLg={8}>

![Vertically stacked versus horizontal alignment](images/checkbox-usage-4.png)

</Column>
</Row>

<Caption>Vertically stacked versus horizontal alignment</Caption>

### Placement

The checkbox component is often used in forms. Forms can be placed on a full page, in a modal or in a side panel. A checkbox can also be used for agreeing to terms and conditions or to filter information.

<Row>
<Column colLg={8}>

![Placement of checkboxes in different UI scenarios.](images/checkbox-usage-5.png)

</Column>
</Row>

> Checkboxes in a form should be placed at least 32px (`layout-03`) below or before the next component. Spacing of 24px or 16px can also be used when space is more restricted or if the form is more complex.

Reviewer 3: do we want to mention the tokens for 24px and 16px here too?

  </DoDont>
</DoDontRow>

### Further guidance

For further content guidance, see Carbon's [content guidelines](notion://www.notion.so/guidelines/content/general).

## Behaviors

### States

The checkbox input allows for a series of states: **unselected**, **selected**, and **indeterminate**. The default view of a set of checkboxes is having no option selected.

<Row>
<Column colLg={8}>

![Unselected, selected and indeterminate states of checkbox.](images/checkbox-usage-7.png)

</Column>
</Row>

Use the indeterminate state when the checkbox contains a sublist of selections, some of which are selected, and some unselected. In addition to unselected, selected, and indeterminate states, checkboxes also have states for focus and disabled.

<Row>
<Column colLg={8}>

![Focus and disabled states of checkbox.](images/checkbox-usage-8.png)

</Column>
</Row>

### Interactions

#### Mouse

Users can trigger an item by clicking on the checkbox input directly or by clicking on the checkbox label. Having both regions interactive creates a more accessible click target.

<Row>
<Column colLg={8}>

![Click targets for checkbox.](images/checkbox-usage-9.png)

</Column>
</Row>

#### Keyboard

> Users can trigger a state change by pressing the `Enter` or `Spacebar` key while the checkbox input has focus. For additional keyboard interactions, see the [accessibility tab](https://www.carbondesignsystem.com/components/checkbox/accessibility).

Reviewer 3: is `Enter` to select/deselect expected behavior? we currently don't have this and it's not default behavior for checkboxes

Author: Okay i will take that out

Reviewer 3: by default, space bar selects and deselects but Enter doesn't, and our components follow that pattern

Reviewer 3: I didn't see your edit originally, but I don't think JAWS and NVDA change the interaction pattern (not entirely sure on that though since I don't have access to either)

Author: @dakahn do you have any insight on this?

Reviewer 4: Space (not enter) to select and deselect is expected default behavior. Unless we have really really good reason to we defer to the [WAI-Aria Authoring Practices Guide](https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190814/examples/checkbox/checkbox-1/checkbox-1.html) for keyboard interaction patterns.

</Column>
</Row>

<Caption>Example of a full page filter list.</Caption>

### When not to use

If a user can select only one option from a list, radio buttons should be used instead of checkboxes. Checkboxes allow the user to select multiple items in a set whereas radio buttons allow the user to select only one option.

<DoDontRow>
  <DoDont caption="Do use radio buttons when only one item can be selected.">

![Do use radio buttons when only one item can be selected.](images/checkbox-usage-2-do.png)

  </DoDont>
  <DoDont type="dont" caption="Don't use checkboxes when only one item can be selected.">

![Don't use checkboxes when only one item can be selected.](images/checkbox-usage-2-dont.png)

  </DoDont>
</DoDontRow>

## Formatting

### Anatomy

The checkbox component is comprised of a label and a checkbox. If there is a group of checkboxes a heading can be added.

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-3.png)

</Column>
</Row>

> 1. **Heading (optional):** Communicates what needs to be selected below.

Reviewer 2: I think we might need to flag this as required potentially, but I think it's right to say that this doesn't have to present itself visually.

In code this group might look like:

```html
<h3 id="group-label">Languages</h3>
<div role="group" aria-labelledby="group-label">
  <ul>
    <li>
      <input id="option-1" type="checkbox" />
      <label for="option-1">Node.js</label>
    </li>
    <li>
      <input id="option-2" type="checkbox" />
      <label for="option-2">Go</label>
    </li>
    <li>
      <input id="option-3" type="checkbox" />
      <label for="option-3">Java</label>
    </li>
  </ul>
</div>
```

And as a result we do need a label for the group, at the very least in code but ideally presented to both user groups.

Author: In code, if there was one checkbox present would you need a heading for it in code? (ex: terms and conditions). if not, then maybe we just need to clarify there that headings for groups of checkboxes is required.

Reviewer 2: @laurenmrice good question, I don't know if we would say heading necessarily but we would need a label. If you look at that markup, we would say that the label for the checkbox is the `<label>` element, for example:

```html
<input id="option-3" type="checkbox" />
<label for="option-3">Java</label>
```

We would say that the label for the checkbox is `Java`.

The group also has a label, but instead, it uses `aria-labelledby` instead of the `<label>` element. The `aria-labelledby` attribute points to the heading `<h3>` tag and treats its text as the label for the group.

Author: Okay, For this we can say something about how visually its not required unless there is a grouping of two or more checkboxes and also say that in code it needs to be required even if its just one checkbox and can reference the code tab for my information.

Author: Is it okay that we are labeling the anatomy of the checkbox as 1. Heading. 2. Checkbox and 3. Label ? If in fact `<label>` is referencing the "heading". In the react demo and on the component switcher demo the text is showing how I am currently labeling them here.

Reviewer 2: @laurenmrice I think it breaks down into:

- A checkbox input requires a label
- A group of checkbox inputs requires a grouping label to describe the group of checkbox inputs
- A label is always required for a checkbox input and group of checkboxes, but in rare cases it does not need to be visible. To understand how to label something without it showing up visually, see ___.

![image](https://user-images.githubusercontent.com/3901764/80030037-4b8d9500-84ad-11ea-9cf0-b8ff5f98e941.png)

Author: In the newest build, I changed all the content on the page to reflect this. ^

2. **Checkbox:** A checkbox icon indicating the appropriate state. By default it is unselected.
3. **Label:** Describes the information you want to select or unselect.

### Alignment

Checkbox labels fall to the right of their controls. If there is a checkbox grouping, they can be laid out vertically or horizontally depending on the use case and the structure of the UI. When possible, arrange the checkbox and radio button groups vertically for easier reading.

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-4.png)

</Column>
</Row>

> <Caption>Vertically stacked versus horizontal alignment.</Caption>

Reviewer: ```suggestion
<Caption>Vertically stacked versus horizontal alignment</Caption>
```

### Placement

> The checkbox component is often used in forms. Forms can be placed on a full page, in a modal or in a side panel. A checkbox can also be used for agreeing to terms and conditions or to filter information.

Reviewer: ```suggestion
The checkbox component is often used in forms. Forms can be placed on a full page, in a modal, or in a side panel. A checkbox can also be used for agreeing to terms and conditions or to filter information.
```

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-5.png)

</Column>
</Row>

> Checkboxes in a form should be placed at least 32px (`layout-03`) below or before the next component. Having 24px or 16px can also be used when space is more restricted or if the form is more complex.

Reviewer: ```suggestion
Checkboxes in a form should be placed at least 32px (`layout-03`) below or before the next component. Spacing of 24px or 16px can also be used when space is more restricted or if the form is more complex.
```

For more information on spacing in forms, see our [form style guidance](/components/form/style).

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-6.png)

</Column>
</Row>

> <Caption>Spacing between a checkbox and other components in a form.</Caption>

Reviewer: ```suggestion
<Caption>Spacing between a checkbox and other components in a form</Caption>
```

For more information about spacing for the checkbox component itself, see the [style tab](/components/checkbox/style).

## Behavior

### States

The checkbox control allows for a series of states: **unselected**, **selected**, and **indeterminate**. The default view of a set of checkboxes is having no option selected.

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-7.png)

</Column>
</Row>

Use the indeterminate state when the checkbox contains a sublist of selections, some of which are selected, and some unselected. In addition to unselected, selected, and indeterminate states, checkboxes also have states for focus and disabled.

<Row>
<Column colLg={8}>

![The default state for checkboxes is unchecked.](images/checkbox-usage-8.png)

</Column>
</Row>

### Interactions

#### Click target

> Users can select an item by clicking on the checkbox control directly or by clicking on the checkbox label. Having both regions interactive creates a more accessible click target.

Reviewer 2: I think the click target area for this might be off by a bit, unfortunately I think the space around the checkbox is margin currently:

![image](https://user-images.githubusercontent.com/3901764/79907568-685b9700-83df-11ea-8c95-c8ce522fa16c.png)

And so wouldn't be interactive 😞 

 I think the click target for the text would be the line-height (24px?). The image could 100% have that exactly, just couldn't tell 😅 

Similarly, if we wanted to do guidance for how to interact with keyboard/SR we could use the same copy from accordion since I believe it's space/enter along with ctrl+option+space!

Author: OH okay, so the click target for the control is just the icon bounds only (not including the margin). And the click target for the label would be the line height.

If you think it would be helpful to add the keyboard guidance we can! Is it the exact same copy from Accordion or does it need to be tweaked?

Reviewer 2: I think it's the same, yeah! The only change would be the focus outline placement.
